### ISSUE TITLE
[discussion] HPO in a CL scenario

### ISSUE THREAD

**Issue by loribonna** [awslabs/Renate#58]
Hi authors, thanks for the awesome library!

I was wondering how the hyperparameter optimization with HPO is implemented in a continual scenario. 
From what I found in your documentation, it seems the stopping criterion is based on `max_time`. From my understanding, this means that the best configuration is chosen as the one with the best `metric` up to the current task, which may result in configurations that have a "slower" learning behavior being dropped in the early stages of training.

For example, if I tend to learn fast on the current tasks I may have a good accuracy at task 2, but this does not necessarily mean that I will have a good accuracy on task 10 due to forgetting.

**Comment by wistuba** [comment]
Hi Lorenzo,

Your understanding is correct, we greedily choose hyperparameters which may lead to suboptimal long-term solutions. The underlying assumption is that we don't know how many tasks we will see.

Do you have a better way to address the problem? I don't think this is an easy task to solve since we don't know what will happen in the future. Will the data distribution shift, how does it shift or will there be any future tasks? While we know the answer to these questions in a controlled experimental setup, we might not know in practical setups.

Best,
Martin

**Comment by loribonna** [comment]
Hi, yes I don't think it is an easy problem to solve, I was just wondering if you had a solution already implemented. 

I guess an initial solution might be to wait for a few tasks to occur and then evaluate the error only on past tasks (if the interest if forgetting). This might be implemented as an additional `metric` which does not consider accuracy on the current task and a delay on `max_time` to let some tasks complete.

Either way, the application of HPO in a CL setting is an interesting problem that is not really explored in the current literature!

Best, 
Lorenzo

**Comment by wistuba** [comment]
Not sure this was clear but `max_time` refers to the time spent on the HPO process when triggering an update.

Whenever the update is triggered, an updated model is expected and therefore will be trained and returned. Therefore, we cannot wait for more tasks to arrive. This must be handled on the user side. However, the user may increase the memory to store more data and achieve a similar behavior to what you have described.

It will certainly be interesting to see some improvements in this setting and we will be happy to improve upon our current solution.

**Comment by loribonna** [comment]
Oh my bad, I did not understand the proper effect of `max_time`. In this scenario, when an update is triggered a new model is trained on the current task or on the whole sequence of tasks? 
In the latter case, I guess also that my original drawback does not occur.

**Comment by wistuba** [comment]
We support both cases. The first is the more practical case where you collect or label data over time and want to update your model on a single task. We refer to this in our documentation simply as a [training job](https://renate.readthedocs.io/en/latest/getting_started/how_to_run_training.html). The second case is what we refer to as [benchmarking](https://renate.readthedocs.io/en/latest/benchmarking/index.html) where you train and evaluate over a sequence of tasks. Latter, is basically a loop going over all tasks to call the function from the first case. Therefore, both have in common that HPO is run per task and not across the entire task sequence. Our intuition was that you cannot tune your hyperparameters on data you have no access to yet. However, you could use few tasks for HPO and then evaluate on the remaining tasks. We do not support this but I think it is interesting and I think we should discuss when we want to address this.

**Comment by 610v4nn1** [comment]
Martin already clarified the point but I would like to make things more concrete to make sure we are on the same page.

When running HPO, it is best practice to tune the hyperparameters measuring the performance on a validation set (e.g., val_accuracy). When updating your model on the i-th chunk of data (or the i-th task), you can have a validation set which contains data from all the chunks from 0 to i. This will automatically account for the impact of the forgetting (assuming the mix in your validation set is representative). Basically, you can just have another memory storing the validation set with something like Reservoir Sampling.

This is still not optimizing over a sequence of tasks, but in practice often happens that each model is deployed to serve predictions for some business. In this case everyone expects each model to be the best possible model respect to what was observed up to that point since otherwise there will be an impact on the quality of the predictions served.
